spade_llm/discovery_platform.py

```diff
@@ -0,0 +1,148 @@
+
+from langchain_chroma import Chroma
+from langchain_core.documents import Document
+from langchain_core.embeddings import Embeddings
+from langchain_core.vectorstores import VectorStore
+from pydantic import BaseModel, Field
+
+from spade_llm.core.agent import Agent
+from spade_llm.demo import models
+
+from spade_llm.core.behaviors import MessageHandlingBehavior, MessageTemplate
+from spade_llm.core.conf import configuration, Configurable
+
+_ID_SEPARATOR = "/"
+
+logger = logging.getLogger(__name__)
+
+
+class AgentTask(BaseModel):
+    """Represents a single task agent is capable for."""
+
+    description: str = Field(description="Description of a task")
+    examples: Optional[list[str]] = Field(description="Examples with a task query")
+
+    def to_string(self) -> str:
+        return self.description + "\n" + "\n".join(self.examples)
+
+
+class AgentDescription(BaseModel):
+    """Represents a description of a certain agent with its functions"""
+
+    description: str = Field(description="Description of the agent role and abilities")
+    tasks: Optional[list[AgentTask]] = Field(description="Tasks agent is capable for", default=[])
+    domain: Optional[str] = Field(description="Domain agent is expert at", default="")
+    tier: int = Field(default=0, description="Tier the agent belongs to. During discovery process agents can "
+                                             "lookup only agents in the tiers below and in the same domain.")
+    id: str = Field(description="Unique identifier of an agent")
+
+    def _to_documents(self) -> list[Document]:
+        yield Document(
+            page_content=self.description,
+            id=_ID_SEPARATOR.join([self.id, '0']),
+            metadata={"domain": self.domain, "tier": self.tier}
+        )
+
+        counter: int = 1
+        for task in self.tasks:
+            counter += 1
+            yield Document(
+                page_content=self.description + "\n" + task.to_string(),
+                id=_ID_SEPARATOR.join([self.id, str(counter)]),
+                metadata={"domain": self.domain, "tier": self.tier}
+            )
+
+    def to_documents(self) -> list[Document]:
+        return list(self._to_documents())
+
+    @staticmethod
+    def extract_agent_id(doc: Document) -> str:
+        return "".join(doc.id.split(_ID_SEPARATOR)[:-1])
+
+
+class AgentSearchRequest(BaseModel):
+    """Search query used to lookup agents for solving tasks"""
+
+    task: str = Field(description="Description of the task we need to find agent for.")
+    top_k: int = Field(default=4, description="Maximum number of agents to report.")
+
+
+class AgentSearchResponse(BaseModel):
+    """Response to the search request with list of the agents"""
+    agents: list[AgentDescription] = Field(description="List of agent descriptions for agents matching search query.")
+
+
+class DirectoryFacilitatorAgentConf(BaseModel):
+    agents: dict = Field(description="Dictionary of agents to register.", default={})
```

Comment:
This is not a part of configuration. Use agent state instead to track registered agents.